architecture/spmc.rst

```diff
@@ -0,0 +1,322 @@
+the normal world. LDELF is used to load the SPs into memory.
+
+Starting SPs
+------------
+SP are loaded and started as the last step in OP-TEEs initialisation process.
+
+.. uml::
+        :width: 800
+
+        autoactivate on
+        thread_optee_smc_a64.s -> boot_final
+        boot_final -> secure_partition.c:	sp_init_all
+        loop for each SP
+        secure_partition.c -> secure_partition.c:	sp_init_uuid
+        return
+        end
+        return
+        return
+        autoactivate off
+        thread_optee_smc_a64.s -> thread_optee_smc_a64.s:	thread_ffa_msg_wait
+        thread_optee_smc_a64.s -> thread_optee_smc_a64.s:	ffa_msg_loop
+        autoactivate off
+        thread_optee_smc_a64.s -> SPMD:		SMC
+
+:sp_init_all:		Initialise all SPs which have been added by the SP_PATHS
+			compiler option and run them
+:thread_ffa_msg_wait:	All SPs are loaded and started. A FF-A wait message is
+			send to the Normal World.
+
+
+Each SP is loaded into the system using ldelf and started. This is done using
+LDELF and is based around the same process as loading the early TAs.
+All SPs are run after they are loaded and run until a FFA_MSG_WAIT is send by
+the SP.
+
+
+.. uml::
+
+        autoactivate on
+        secure_partition.c -> secure_partition.c:	sp_init_uuid
+        secure_partition.c -> secure_partition.c:	sp_open_session
+        secure_partition.c -> secure_partition.c:	find_sp
+        return
+        secure_partition.c -> secure_partition.c:	sp_create_session
+        return
+        secure_partition.c -> secure_partition.c:	ldelf_load_ldelf
+        return
+        secure_partition.c -> secure_partition.c:	ldlelf_init_with_ldelf
+        return
+        secure_partition.c -> secure_partition.c:	sp_init_set_registers
+        return
+        return
+        secure_partition.c -> secure_partition.c:	enter_sp
+        return
+        secure_partition.c -> secure_partition.c:	sp_msg_handler
+        return
+        return
+
+:init_with_ldelf: Load the SP
+:sp_init_info:	Initialise the ``ffa_init_info``. The ``ffa_init_info`` is
+		passed to the SP during it first run.
+:sp_init_set_registers: Initialise the registers of the SP
+:sp_msg_handler:	Handle the SPs FF-A message
+
+Once all SP are loaded and started we return to the SPMD and the Normal World is
+booted.
+
+
+SP message handling
+-------------------
+
+The SPMC is split into 2 main message handlers:
+
+:thread_spmc_msg_recv() thread_spm.c:	used to handle message coming from the
+					Normal World.
+:sp_msg_handler() spmc_sp_handler.c:	Used to handle message where the source
+					or the destination is a SP.
+
+When a FFA_MSG_SEND_DIRECT_REQ message is received by the spmc from the Normal
+World, a new kernel thread is started.
+The FF-A message is passed to the kernel thread and it will call the
+sp_msg_handler function.
+
+Whenever the SPMC (sp_msg_handler) receives a message not intended for one of
```

Comment:
Shouldn't all function names be in double back ticks and have parentheses?
```
``sp_msg_handler()``
```

Reply:
Done